File: packages/java-pom-bdd/src/test/java/com/example/pages/LoginPage.java
package com.example.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class LoginPage extends BasePage {
    private final Locator dialog;
    private final Locator usernameInput;
    private final Locator passwordInput;
    private final Locator loginButton;
    private final Locator registerLink;

    public LoginPage(Page page) {
        super(page);
        this.dialog = page.getByRole(com.microsoft.playwright.options.AriaRole.DIALOG,
                new Page.GetByRoleOptions().setName("ログイン"));
        this.usernameInput = dialog.getByLabel("ユーザー名:");
        this.passwordInput = dialog.getByLabel("パスワード:");
        this.loginButton = dialog.getByRole(com.microsoft.playwright.options.AriaRole.BUTTON,
                new Locator.GetByRoleOptions().setName("ログイン"));
        this.registerLink = dialog.getByRole(com.microsoft.playwright.options.AriaRole.LINK,
                new Locator.GetByRoleOptions().setName("新規登録"));
    }

    public boolean isVisible() {
        return dialog.isVisible();
    }

    public void login(String username, String password) {
        usernameInput.fill(username);
        passwordInput.fill(password);
        loginButton.click();
    }

    public void clickRegisterLink() {
        registerLink.click();
    }
}

File: packages/ts-pom-bdd-di/pages/LoginPage.ts
import { Page, Locator } from '@playwright/test';

export class LoginPage {
  private readonly dialog: Locator;
  private readonly usernameInput: Locator;
  private readonly passwordInput: Locator;
  private readonly loginButton: Locator;
  private readonly closeButton: Locator;
  private readonly registerLink: Locator;
  private readonly errorMessage: Locator;

  constructor(private page: Page) {
    this.dialog = page.getByRole('dialog', { name: 'ログイン' });
    this.usernameInput = this.dialog.getByLabel('ユーザー名:');
    this.passwordInput = this.dialog.getByLabel('パスワード:');
    this.loginButton = this.dialog.getByRole('button', { name: 'ログイン' });
    this.closeButton = this.dialog.getByRole('button', { name: '閉じる' });
    this.registerLink = this.dialog.getByRole('link', { name: '新規登録' });
    this.errorMessage = this.dialog.locator('[class*="error"]');
  }

  async isVisible() {
    return await this.dialog.isVisible();
  }

  async login(username: string, password: string) {
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.loginButton.click();
  }

  async close() {
    await this.closeButton.click();
  }

  async clickRegisterLink() {
    await this.registerLink.click();
  }

  async getErrorMessage() {
    if (await this.errorMessage.isVisible()) {
      return await this.errorMessage.textContent();
    }
    return null;
  }

  async hasError() {
    return await this.errorMessage.isVisible();
  }
}
